### ISSUE TITLE
more than 1 promise `.catch()` causing query to be executed more than once

### ISSUE THREAD

**Issue by unlucio** [tgriesser/knex#2726]
# Environment

Knex version:
from knex=0.12.9 + pg=6.1.2
to knex=0.15.2 + pg=7.4.3 (included)

OS: osx, windows, linux

Select applicable tempalate from below.
If issue is about oracledb support tag @ atiertant. For MSSql tag @ smorey2 .
Rest of dialects doesn't need tags.

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do
Chaining multiple `.catch()` traps, for a total of Ncatch, to `insert()` causes the insert to be repeated for (Ncatch -1) + 1 times (the -1 being the catch it skips, and +1 been the expected one we actually created).
Expected behaviour: chaining multiple catches to a promise interface should behave like a promise without weird side effetcs to the query.

2. Error message
If the target table has constrain than we'll have a duplicate insert error, otherwise we'll just find multiple entries in out tables without any evidence of what caused it.

3. Reduced test code, for example in https://npm.runkit.com/knex or if it needs real
   database connection to mysql or postgresql, then single file example which initializes
   needed data and demonstrates the problem.

Consider the following code:
```javascript
const knex = require('knex');

const cred = {
    dialect: 'postgres',
    debug: true,
    connection:{
        host: 'localhost',
        port: '5432',
        user: 'docker',
        password: 'docker',
        database: 'c8_accounts',
        ssl: false
    },
    pool: { min: 5, max: 25 }
};

const conn = knex(cred);
const tableName = 'clients';
const data = { name: `whatever-${Date.now()}`, status: 'whatever-test' };

const writer = conn(tableName).insert(data);
writer.then((result) => {
    console.log(`add record to ${tableName} (%o) result: `, data, result);
    return result;
}).catch((error) => {
        console.error(`add record to ${tableName} (%o) error: `, data, error);
    return Promise.reject(error);
});
```
This has the expect output (`debug: true`):
```bash
{ method: 'insert',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 'whatever-1532357467509', 'whatever-test' ],
  __knexQueryUid: '5a5cdebd-7b09-43e0-9b96-fa1f7aff9348',
  sql: 'insert into "clients" ("name", "status") values (?, ?)',
  returning: undefined }

add record to clients ({ name: 'whatever-1532357467509', status: 'whatever-test' }) result:  Result {
  command: 'INSERT',
  rowCount: 1,
  oid: 0,
  rows: [],
  fields: [],
  _parsers: [],
  RowCtor: null,
  rowAsArray: false,
  _getTypeParser: [Function: bound ] }
```

Now if we add another catch statement (ex: `writer.catch((error) => console.log('just because: ', error))`):
```javascript
const knex = require('knex');

const cred = {
    dialect: 'postgres',
    debug: true,
    connection:{
        host: 'localhost',
        port: '5432',
        user: 'docker',
        password: 'docker',
        database: 'c8_accounts',
        ssl: false
    },
    pool: { min: 5, max: 25 }
};

const conn = knex(cred);
const tableName = 'clients';
const data = { name: `whatever-${Date.now()}`, status: 'whatever-test' };

const writer = conn(tableName).insert(data);
writer.then((result) => {
    console.log(`add record to ${tableName} (%o) result: `, data, result);
    return result;
}).catch((error) => {
        console.error(`add record to ${tableName} (%o) error: `, data, error);
    return Promise.reject(error);
});

writer.catch((error) => console.log('just because: ', error))
```
Suddenly something unexpected happens:
```javascript
{ method: 'insert',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 'whatever-1532357719493', 'whatever-test' ],
  __knexQueryUid: 'f7ced535-6b3c-440a-8e37-527a6453bc72',
  sql: 'insert into "clients" ("name", "status") values (?, ?)',
  returning: undefined }
{ method: 'insert',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 'whatever-1532357719493', 'whatever-test' ],
  __knexQueryUid: '9edfa735-37e6-4684-ae72-4930045eefa2',
  sql: 'insert into "clients" ("name", "status") values (?, ?)',
  returning: undefined }
```
the knex-provided debug output has now 2 jobs to do (with different __knexQueryUid, so they are actually different things)

and we get both the success log line:
```bash
add record to clients ({ name: 'whatever-1532357719493', status: 'whatever-test' }) result:  Result {
  command: 'INSERT',
  rowCount: 1,
  oid: 0,
  rows: [],
  fields: [],
  _parsers: [],
  RowCtor: null,
  rowAsArray: false,
  _getTypeParser: [Function: bound ] }
```

as well as the __2nd__ catch output (please note the __1st catch was ignored__):
```bash
just because:  { error: duplicate key value violates unique constraint "clients_name_key"
    at Connection.parseE (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:553:11)
    at Connection.parseMessage (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:378:19)
    at Socket.<anonymous> (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:119:22)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:277:12)
    at readableAddChunk (_stream_readable.js:262:11)
    at Socket.Readable.push (_stream_readable.js:217:10)
    at TCP.onread (net.js:638:20)
  name: 'error',
  length: 217,
  severity: 'ERROR',
  code: '23505',
  detail: 'Key (name)=(whatever-1532357719493) already exists.',
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: 'public',
  table: 'clients',
  column: undefined,
  dataType: undefined,
  constraint: 'clients_name_key',
  file: 'nbtinsert.c',
  line: '433',
  routine: '_bt_check_unique' }
```

The more `catch()` we add:
```javascript
const knex = require('knex');

const cred = {
    dialect: 'postgres',
    debug: true,
    connection:{
        host: 'localhost',
        port: '5432',
        user: 'docker',
        password: 'docker',
        database: 'c8_accounts',
        ssl: false
    },
    pool: { min: 5, max: 25 }
};

const conn = knex(cred);
const tableName = 'clients';
const data = { name: `whatever-${Date.now()}`, status: 'whatever-test' };

const writer = conn(tableName).insert(data);
writer.then((result) => {
    console.log(`add record to ${tableName} (%o) result: `, data, result);
    return result;
}).catch((error) => {
        console.error(`add record to ${tableName} (%o) error: `, data, error);
    return Promise.reject(error);
});

writer.catch((error) => console.log('just because1: ', error))
writer.catch((error) => console.log('just because2: ', error))
writer.catch((error) => console.log('just because3: ', error))
writer.catch((error) => console.log('just because4: ', error))
```

the more times knex tries to execute our query 5 times (4 additional `catches` + our `insert`):
```javascript
{ method: 'insert',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 'whatever-1532358106864', 'whatever-test' ],
  __knexQueryUid: '3483dba3-c371-48bc-a6f2-0528ae7f5805',
  sql: 'insert into "clients" ("name", "status") values (?, ?)',
  returning: undefined }
{ method: 'insert',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 'whatever-1532358106864', 'whatever-test' ],
  __knexQueryUid: '8657dbeb-f192-47e5-a511-aed25f7af91e',
  sql: 'insert into "clients" ("name", "status") values (?, ?)',
  returning: undefined }
{ method: 'insert',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 'whatever-1532358106864', 'whatever-test' ],
  __knexQueryUid: '58eee2a1-e39c-486c-a563-23c62eb95a73',
  sql: 'insert into "clients" ("name", "status") values (?, ?)',
  returning: undefined }
{ method: 'insert',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 'whatever-1532358106864', 'whatever-test' ],
  __knexQueryUid: '185e8e27-e5f2-482c-a12d-fdd376a5dd32',
  sql: 'insert into "clients" ("name", "status") values (?, ?)',
  returning: undefined }
{ method: 'insert',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 'whatever-1532358106864', 'whatever-test' ],
  __knexQueryUid: '21945a1d-0b35-45c4-a1c9-dd2be76e68b5',
  sql: 'insert into "clients" ("name", "status") values (?, ?)',
  returning: undefined }
```
(please not how all the `__knexQueryUid` are different)

and we still get our success:
```bash
add record to clients ({ name: 'whatever-1532358106864', status: 'whatever-test' }) result:  Result {
  command: 'INSERT',
  rowCount: 1,
  oid: 0,
  rows: [],
  fields: [],
  _parsers: [],
  RowCtor: null,
  rowAsArray: false,
  _getTypeParser: [Function: bound ] }
```
as well as a (totalCatches)-1 number of failure outputs:
```bash
just because2:  { error: duplicate key value violates unique constraint "clients_name_key"
    at Connection.parseE (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:553:11)
    at Connection.parseMessage (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:378:19)
    at Socket.<anonymous> (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:119:22)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:277:12)
    at readableAddChunk (_stream_readable.js:262:11)
    at Socket.Readable.push (_stream_readable.js:217:10)
    at TCP.onread (net.js:638:20)
  name: 'error',
  length: 217,
  severity: 'ERROR',
  code: '23505',
  detail: 'Key (name)=(whatever-1532358106864) already exists.',
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: 'public',
  table: 'clients',
  column: undefined,
  dataType: undefined,
  constraint: 'clients_name_key',
  file: 'nbtinsert.c',
  line: '433',
  routine: '_bt_check_unique' }
just because4:  { error: duplicate key value violates unique constraint "clients_name_key"
    at Connection.parseE (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:553:11)
    at Connection.parseMessage (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:378:19)
    at Socket.<anonymous> (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:119:22)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:277:12)
    at readableAddChunk (_stream_readable.js:262:11)
    at Socket.Readable.push (_stream_readable.js:217:10)
    at TCP.onread (net.js:638:20)
  name: 'error',
  length: 217,
  severity: 'ERROR',
  code: '23505',
  detail: 'Key (name)=(whatever-1532358106864) already exists.',
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: 'public',
  table: 'clients',
  column: undefined,
  dataType: undefined,
  constraint: 'clients_name_key',
  file: 'nbtinsert.c',
  line: '433',
  routine: '_bt_check_unique' }
just because3:  { error: duplicate key value violates unique constraint "clients_name_key"
    at Connection.parseE (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:553:11)
    at Connection.parseMessage (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:378:19)
    at Socket.<anonymous> (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:119:22)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:277:12)
    at readableAddChunk (_stream_readable.js:262:11)
    at Socket.Readable.push (_stream_readable.js:217:10)
    at TCP.onread (net.js:638:20)
  name: 'error',
  length: 217,
  severity: 'ERROR',
  code: '23505',
  detail: 'Key (name)=(whatever-1532358106864) already exists.',
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: 'public',
  table: 'clients',
  column: undefined,
  dataType: undefined,
  constraint: 'clients_name_key',
  file: 'nbtinsert.c',
  line: '433',
  routine: '_bt_check_unique' }
just because1:  { error: duplicate key value violates unique constraint "clients_name_key"
    at Connection.parseE (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:553:11)
    at Connection.parseMessage (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:378:19)
    at Socket.<anonymous> (/Users/lucio/tmp/dbexp/node_modules/pg/lib/connection.js:119:22)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:277:12)
    at readableAddChunk (_stream_readable.js:262:11)
    at Socket.Readable.push (_stream_readable.js:217:10)
    at TCP.onread (net.js:638:20)
  name: 'error',
  length: 217,
  severity: 'ERROR',
  code: '23505',
  detail: 'Key (name)=(whatever-1532358106864) already exists.',
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: 'public',
  table: 'clients',
  column: undefined,
  dataType: undefined,
  constraint: 'clients_name_key',
  file: 'nbtinsert.c',
  line: '433',
  routine: '_bt_check_unique' }
```

This erratic behaviour can be prevented by simply coercing `insert`'s to a canonical native Promise:
```javascript
const knex = require('knex');

const cred = {
    dialect: 'postgres',
    debug: true,
    connection:{
        host: 'localhost',
        port: '5432',
        user: 'docker',
        password: 'docker',
        database: 'c8_accounts',
        ssl: false
    },
    pool: { min: 5, max: 25 }
};

const conn = knex(cred);
const tableName = 'clients';
const data = { name: `whatever-${Date.now()}`, status: 'whatever-test' };

const writer = new Promise(function (resolve, reject) {
    conn(tableName).insert(data).then(resolve).catch(reject);
});

writer.then((result) => {
    console.log(`add record to ${tableName} (%o) result: `, data, result);
    return result;
}).catch((error) => {
        console.error(`add record to ${tableName} (%o) error: `, data, error);
    return Promise.reject(error);
});

writer.catch((error) => console.log('just because1: ', error))
writer.catch((error) => console.log('just because2: ', error))
writer.catch((error) => console.log('just because3: ', error))
writer.catch((error) => console.log('just because4: ', error))
```

and we're back to an expected output:
```bash
{ method: 'insert',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 'whatever-1532358850252', 'whatever-test' ],
  __knexQueryUid: '5b781898-eee5-4fdf-a11d-c9011cc1d4c3',
  sql: 'insert into "clients" ("name", "status") values (?, ?)',
  returning: undefined }
add record to clients ({ name: 'whatever-1532358850252', status: 'whatever-test' }) result:  Result {
  command: 'INSERT',
  rowCount: 1,
  oid: 0,
  rows: [],
  fields: [],
  _parsers: [],
  RowCtor: null,
  rowAsArray: false,
  _getTypeParser: [Function: bound ] }
```

Long story short, this is pretty much like having something simple like:
```
function executeInsert(table, data) {
    return conn(table).insert(data)
    .catch((error) => {
        console.error('DB ERROR: ', error);
        return Promise.reject(error);
    });
}

function addSomething(data) {
    return executeInsert('myTable', data)
        .catch(function(error) {
            console.error('Cannot add something. Error: ', error);
        })
}

addSomething({name: 'nothing'}).then(function (result) {
    console.log('Somehing was inserted: ', result);
})
```
unexpectedly adding something a number of times.

**Comment by ricardograca** [comment]
When you call `.catch()` on a Promise you're not "adding" a generic error handler that will only run when there is an error or something, you're actually running this: `.then(null, function() { /* ... */ })`, so what you are seeing is perfectly expected since you're running multiple queries in parallel, each one started when you call `.catch()` on the insert Promise.

What you really want is this:

```js
writer.then((result) => {
    console.log(`add record to ${tableName} (%o) result: `, data, result);
    return result;
}).catch((error) => {
        console.error(`add record to ${tableName} (%o) error: `, data, error);
    return Promise.reject(error);
}).catch((error) => console.log('just because: ', error))
```

You can chain `.catch()` calls to eternity like that and only one query will run, but if you call `.catch()` on `writer` it will simply start another Promise chain in parallel with the first one.

**Comment by unlucio** [comment]
@ricardograca thanks for your answer, however catching and throwing between `catches` in a promise is something I use quite a bit (usually for logging purposes) and I never observed a behavior like this.
Also: how do you explain wrapping it in a "js native" promise fixes the problem?

**Comment by ricardograca** [comment]
Not really sure, but maybe it's related to the use of Bluebird in Knex instead of using native Promises.

**Comment by wubzz** [comment]
Unless I'm mistaken this is primarily because knex interface is a "thenable", not a "real promise" if you will. For example if you call `.then` on a query builder multiple times it may not always yield the same result, since it actually does run the query every time its called, unlike a "real promise" which would immediately resolve with the original result.

See [here](https://github.com/tgriesser/knex/blob/master/src/interface.js#L14) and [here](https://github.com/tgriesser/knex/blob/master/src/interface.js#L107).

It has been lifted as an issue multiple times before. I believe in @tgriesser 1.0 refactor branch this was [changed](https://github.com/wubzz/knex/commit/d8852bb8038f35b0e810465dab2b4dbbfec35eda#diff-ff9eb136561ea616d243650733d24677R78) to more accurately fulfill the promise spec, right?

**Comment by elhigu** [comment]
Yep, you are calling then / catch for query builder multiple times. Its pretty much completely by valid functionality according to Promise A+ specs for thenable. If that behaviour is changed it will actually make using `await` with knex harder, because you wouldn't be able to execute the same query builder multiple times by awaiting it. There are many issues about this where all the effect are explained better for example here https://github.com/tgriesser/knex/issues/1952#issuecomment-285636299.

I really hope that this will not be changed for 1.0 release. Closing as duplicate.

**Comment by elhigu** [comment]
Some related issue also https://github.com/tgriesser/knex/issues/2434
